### ISSUE TITLE
First published dates in the future

### ISSUE THREAD

**Issue by andylolz** [IATI/ckanext-iati#262]
The following publishers all have first published dates in the future (which shouldn’t be possible)

 * https://iatiregistry.org/publisher/oxford - 01 August 2019
 * https://iatiregistry.org/publisher/aadk - 02 August 2019
 * https://iatiregistry.org/publisher/zoa_uk - 04 August 2019
 * https://iatiregistry.org/publisher/globalcanopy - 05 August 2019
 * https://iatiregistry.org/publisher/liwa-admin - 05 August 2019
 * [https://iatiregistry.org/publisher/silsab1-_](https://iatiregistry.org/publisher/silsab1-_) - 05 August 2019
 * https://iatiregistry.org/publisher/missioneast - 01 September 2019
 * https://iatiregistry.org/publisher/rmf - 04 September 2019
 * https://iatiregistry.org/publisher/shn - 04 September 2019
 * https://iatiregistry.org/publisher/dhrc_nepal - 04 October 2019
 * https://iatiregistry.org/publisher/umu - 05 October 2019
 * https://iatiregistry.org/publisher/acatba - 01 November 2019
 * https://iatiregistry.org/publisher/universiteit_leiden - 02 November 2019
 * [https://iatiregistry.org/publisher/\_path\_](https://iatiregistry.org/publisher/_path_) - 03 November 2019
 * https://iatiregistry.org/publisher/dpod - 03 December 2019
 * https://iatiregistry.org/publisher/ipe - 03 December 2019
 * https://iatiregistry.org/publisher/rainforest-alliance - 03 December 2019

**Comment by andylolz** [comment]
In many cases, this seems to be a result of muddling US and non-US date ordering e.g.

Publisher | First published date | Actual first published date
----------------------------------------------|------------|-----------
[fyf](https://iatiregistry.org/publisher/fyf) | 2012-07-06 | 2012-06-07
[ffi_publisher](https://iatiregistry.org/publisher/ffi_publisher) | 2012-07-06 | 2012-06-07
[basicneeds](https://iatiregistry.org/publisher/basicneeds) | 2012-11-06 | 2012-06-11
[buildafrica](https://iatiregistry.org/publisher/buildafrica) | 2012-12-06 | 2012-06-12
[power](https://iatiregistry.org/publisher/power) | 2012-11-07 | 2012-07-11
[theict](https://iatiregistry.org/publisher/theict) | 2012-11-07 | 2012-07-11
[vfan](https://iatiregistry.org/publisher/vfan) | 2012-12-11 | 2012-11-12
[unitedstates](https://iatiregistry.org/publisher/unitedstates) | 2013-02-01 | 2013-01-02
[wwfpakistan](https://iatiregistry.org/publisher/wwfpakistan) | 2013-04-03 | 2013-03-04
[iadb](https://iatiregistry.org/publisher/iadb) | 2013-08-03 | 2013-03-08
[gain](https://iatiregistry.org/publisher/gain) | 2013-07-05 | 2013-05-07
[goal](https://iatiregistry.org/publisher/goal) | 2013-10-05 | 2013-05-10
[ff](https://iatiregistry.org/publisher/ff) | 2013-11-07 | 2013-07-11
[dc](https://iatiregistry.org/publisher/dc) | 2013-10-09 | 2013-09-10
[blueventures](https://iatiregistry.org/publisher/blueventures) | 2013-11-09 | 2013-09-11
[sacbc](https://iatiregistry.org/publisher/sacbc) | 2013-12-09 | 2013-09-12
[cal](https://iatiregistry.org/publisher/cal) | 2014-03-01 | 2014-01-03
[france](https://iatiregistry.org/publisher/france) | 2014-03-01 | 2014-01-03
[drf](https://iatiregistry.org/publisher/drf) | 2014-09-01 | 2014-01-09
[starsfoundation](https://iatiregistry.org/publisher/starsfoundation) | 2014-05-02 | 2014-02-05
[base_22](https://iatiregistry.org/publisher/base_22) | 2014-10-02 | 2014-02-10
[wcf](https://iatiregistry.org/publisher/wcf) | 2014-05-03 | 2014-03-05
[viva](https://iatiregistry.org/publisher/viva) | 2014-05-03 | 2014-03-05
[plan_ireland](https://iatiregistry.org/publisher/plan_ireland) | 2014-05-03 | 2014-03-05

**Comment by andylolz** [comment]
In a few cases, the first published date is wrong by quite a long way:

Publisher | First published date | Actual first published date
--------|------------|-----------
[oxfamgb](https://iatiregistry.org/publisher/oxfamgb) | 2011-11-25 | 2012-08-08
[asdb](https://iatiregistry.org/publisher/asdb) | 2011-11-28 | 2013-06-27
[pwyf](https://iatiregistry.org/publisher/pwyf) | 2011-11-29 | 2017-11-01
[troc](https://iatiregistry.org/publisher/troc) | 2012-05-31 | 2019-04-03
[spark](https://iatiregistry.org/publisher/spark) | 2014-04-28 | 2015-09-30
[rojiroti](https://iatiregistry.org/publisher/rojiroti) | 2014-06-02 | 2019-06-12
[aiddata](https://iatiregistry.org/publisher/aiddata) | 2015-10-14 | 2019-06-12
[wygi](https://iatiregistry.org/publisher/wygi) | 2015-12-23 | 2019-06-10
[gnpplus](https://iatiregistry.org/publisher/gnpplus) | 2017-01-27 | 2018-10-02

**Comment by andylolz** [comment]
Also, the contents of the `publisher_first_publish_date` metadata uses three different formats. It would be good to unify these.

**Comment by gtkChop** [comment]
@andylolz @amy-silcock 
Thanks for bringing this to our notice. Yeah, last time I have updated all publisher date using a script, where I have used python dateutil library caused a mismatch while reading from excel date.  I believe this is not because of IATI system. 

@amy-silcock 
Could you please forward the latest first published date list? I will update in the staging and then move to prod.

Thanks

**Comment by gtkChop** [comment]
@andylolz @amy-silcock 

- [ ] I have updated the first published date as per given csv file. 
- [ ] Changed the date format to standard ISO format. 
- [ ] Template changes to reflect new date format.
- [ ] Added server-side validation for first_published date (earlier there was no server-side validation for the acceptable date format).
- [ ] Metadata is standardized. Now publisher date should be in the format - ```2018-08-09T00:00:00.000000```

@amy-silcock Could you please verify from your side? Please note - **form First Published date help text and date format is changed**

**Comment by gtkChop** [comment]
@amy-silcock Any update on this? If this looks good, I will deploy on production. Thanks

**Comment by amy-silcock** [comment]
Thanks @gtkChop this looks good to us. Please go ahead and deploy to production.

**Comment by gtkChop** [comment]
Updated dates in production and standardized first published date to ISO format. Please let me know if face any issues. Thanks

**Comment by andylolz** [comment]
Yep – same for Asian Development Bank: https://iatiregistry.org/publisher/asdb

**Comment by amy-silcock** [comment]
@andylolz we are using our internal capture of first published dates. Not sure where you're getting your list from? I imagine there are edge cases where an organisation once published in error, or only published a file containing their name. This means the datastore capture of first published dates will in some places differ from our internal list.

**Comment by andylolz** [comment]
> Not sure where you're getting your list from?

I’m using the `metadata_created` date from the registry API, e.g.:
https://iatiregistry.org/api/3/action/package_show?id=asdb-org

![Screenshot 2019-08-07 at 12 19 55](https://user-images.githubusercontent.com/464193/62619048-af673200-b90d-11e9-9c18-23c5e48a3fde.png)

**Comment by andylolz** [comment]
Additionally, there’s a handful of publishers that still use a weird date format for their `publisher_first_publish_date` metadata.

This now shows on the registry as “Date is not valid”:
![Screenshot 2019-08-07 at 12 48 04](https://user-images.githubusercontent.com/464193/62620595-a4160580-b911-11e9-9a85-a85bdd88f13a.png)

These are:

 * [Netherlands - Ministry of Foreign Affairs](https://iatiregistry.org/publisher/minbuza_nl) (19.09.2011)
 * [Stichting Humana](https://iatiregistry.org/publisher/humana_houben) (23.01.2015)
 * [Tearfund](https://iatiregistry.org/publisher/tearfund) (18.07.2012)
 * [SOS Children's Villages The Netherlands](https://iatiregistry.org/publisher/sosnl) (07.06.2017)
 * [OpenNorth](https://iatiregistry.org/publisher/opennorth) (22.02.2019)
 * [IDinsight](https://iatiregistry.org/publisher/idinsight) (27.03.2019)
 * [International Aids Vaccine Initiative (IAVI)](https://iatiregistry.org/publisher/iavi) (22.07.2019)
 * [International Council of Voluntary Agencies](https://iatiregistry.org/publisher/icva) (26.07.2019)

**Comment by gtkChop** [comment]
These are the dates which were not updated by the script.
 For e.g. Netherlands - Ministry of Foreign Affairs - the update was failed because IATI-org identifier already exists.

**Comment by gtkChop** [comment]
@andylolz @amy-silcock 

Multiple same (duplicate) IATI active publisher_id for a different organization. Hence several updates failed.

Query here:
```
select key, value, state, count(*) from group_extra where key='publisher_iati_id' and state='active' group by key, value, state having count(*) > 1;
```
In total there are 52 publisher_id duplicates. I can provide you the list by email.

**Comment by andylolz** [comment]
[Stichting Humana](https://iatiregistry.org/publisher/humana_houben) and [Sympany+](https://iatiregistry.org/publisher/simone_sympany) appear to have the same publisher IATI ID… I didn’t find any others…

**Comment by gtkChop** [comment]
We are running a script to update all publisher_first_published_date. Which utilizes the API action called organization_patch. If any of the validation fails (e.g. in this case organization id already exists), organization patch will fail. Hence first_published_date is not updated.

**Comment by andylolz** [comment]
Okay, I see – thanks for clarifying, @gtkChop.

I checked again and wasn’t able to find publishers using the same `publisher_iati_id` (other than [the one example I gave here](https://github.com/IATI/ckanext-iati/issues/262#issuecomment-520421024)). Are you able to provide an example? E.g. who is the other publisher using XM-DAC-7 (other than [Netherlands - Ministry of Foreign Affairs](https://iatiregistry.org/publisher/minbuza_nl))

**Comment by gtkChop** [comment]
@andylolz @amy-silcock 

Update failed - i.e. org_id already exists is from the validator. This is also related to one of the issue raised in zendesk.

Validator: https://github.com/IATI/ckanext-iati/blob/master/ckanext/iati/logic/validators.py#L178

There are two orgs with value for XM-DAC-7.

Validator Query: This fails because multiple publishers has the same value. (bug)

```
publisher_id_exists = session.query(model.Group)\
        .join((model.GroupExtra, model.Group.id == model.GroupExtra.group_id))\
        .filter(model.GroupExtra.value == publisher_iati_id).first()
```

```
select * from public.group, group_extra where public.group.id=group_extra.group_id and group_extra.value ='XM-DAC-7';
```

```
                  id                  |               group_id               |        key         |  value   | state  |             revision_id              
--------------------------------------+--------------------------------------+--------------------+----------+--------+--------------------------------------
 1566caec-c845-469f-98d1-9f7eb8ee9655 | 7290642a-2d07-43e1-a4c8-f0d1c24eb5a2 | publisher_agencies | XM-DAC-7 | active | fbe9dc3c-a504-46e9-9f92-f9e887415220
 18ae6647-9232-4756-92f3-bb8551f6d7f5 | 82e48ba1-1e13-43ca-9106-d6b18eeea5e8 | publisher_iati_id  | XM-DAC-7 | active | 7d7d442b-5db9-4d0d-a5a2-7c18bb7715
```

I think the validator query must be changed to 
```
publisher_id_exists = session.query(model.Group)\
        .join((model.GroupExtra, model.Group.id == model.GroupExtra.group_id))\
        .filter(and_ (model.GroupExtra.value == publisher_iati_id, model.GroupExtra.key == 'publisher_iati_id')).first()
```

**Comment by gtkChop** [comment]
Need manual intervention for these 12 publishers:

```
 cb56f265-7a32-4e68-a008-0614eb61093e
 953730fe-9a4a-42d1-a30e-d3e70ca35446
 953730fe-9a4a-42d1-a30e-d3e70ca35446
 953730fe-9a4a-42d1-a30e-d3e70ca35446
 953730fe-9a4a-42d1-a30e-d3e70ca35446
 dd9b8d0c-99d0-4922-8d1e-35447c0026cc
 bc99afcd-b4cc-4bfa-a37e-e1dfd5ede7a7
 92e76d45-fd97-4e3f-8867-3547dd8c2249
 a09d19c1-21c2-462f-80a3-df1eb6e0d02f
 64a2c335-5abf-4a88-ba46-f93348a5825e
 1765a07c-f75c-488d-a7e0-4023224f2b1d
 98afbe00-85b5-4072-ba0d-c043119748f2
```

**Comment by gtkChop** [comment]
Below are the statistics:

Empty published date: 10 Organization (Needs manual update)

```
                org_id                | package_count |             key              | value 
--------------------------------------+---------------+------------------------------+-------
 cb56f265-7a32-4e68-a008-0614eb61093e |             1 | publisher_first_publish_date | 
 953730fe-9a4a-42d1-a30e-d3e70ca35446 |             4 | publisher_first_publish_date | 
 dd9b8d0c-99d0-4922-8d1e-35447c0026cc |             1 | publisher_first_publish_date | 
 37827240-cf54-4800-b6d7-70209374f4b7 |             1 | publisher_first_publish_date | 
 bc99afcd-b4cc-4bfa-a37e-e1dfd5ede7a7 |             1 | publisher_first_publish_date | 
 92e76d45-fd97-4e3f-8867-3547dd8c2249 |             1 | publisher_first_publish_date | 
 a09d19c1-21c2-462f-80a3-df1eb6e0d02f |             1 | publisher_first_publish_date | 
 64a2c335-5abf-4a88-ba46-f93348a5825e |             1 | publisher_first_publish_date | 
 1765a07c-f75c-488d-a7e0-4023224f2b1d |             1 | publisher_first_publish_date | 
 98afbe00-85b5-4072-ba0d-c043119748f2 |             1 | publisher_first_publish_date | 
```

Publisher whose published date is still in old format: (10 Publisher)

```
 ab1b67e1-d303-466a-bad2-7620c4601717 |             2 | publisher_first_publish_date | 18.07.2012 |     10
 b9be6bd5-3610-4cff-8dff-25e88ce4a979 |             1 | publisher_first_publish_date | 27.06.2014 |     10
 09e09829-bf18-4ae3-be01-620643659cd5 |             2 | publisher_first_publish_date | 23.01.2015 |     10
 6345ff17-aed0-466e-b081-3c468c24f2af |             6 | publisher_first_publish_date | 07.06.2017 |     10
 a6007398-1ff6-4af8-aa37-0310b58a270f |             2 | publisher_first_publish_date | 13.02.2017 |     10
 fe9e6bcc-a3da-4498-921d-ee2af2fda8cd |             2 | publisher_first_publish_date | 22.02.2019 |     10
 8dcb9523-1470-4997-b28f-91ce7d041aa7 |             2 | publisher_first_publish_date | 27.03.2019 |     10
 000df420-682b-4ec7-934c-5e883cd72e22 |             1 | publisher_first_publish_date | 16.07.2019 |     10
 eae45e0e-6245-4cbe-aff7-080c4ffbfb6b |             2 | publisher_first_publish_date | 22.07.2019 |     10
 bab95cd5-983a-4e42-b7af-f05e6f9b6cdd |             1 | publisher_first_publish_date | 26.07.2019 |     10
```


All the above organization failed to update by the script because of the validation error. Hope this helps to rectify this issue.


Query1:
```
select org_id, cnt as package_count, key, value from (select org.id as org_id, count(*) as cnt from (select id, state from public.group where public.group.state = 'active' and public.group.is_organization=True) as org,(select id as package_id, owner_org from package) as package where org.id = package.owner_org group by org_id) as tb, group_extra where tb.org_id=group_id and key='publisher_first_publish_date' and value='';
```

Query2:
```
select org_id, cnt as package_count, key, value, length(value) from (select org.id as org_id, count(*) as cnt from (select id, state from public.group where public.group.state = 'active' and public.group.is_organization=True) as org,(select id as package_id, owner_org from package) as package where org.id = package.owner_org group by org_id) as tb, group_extra where tb.org_id=group_id and key='publisher_first_publish_date' and value !='' and length(value)<26;
```

**Comment by JoshStanleyDevInit** [comment]
Thanks for providing those @gtkChop 

I have attached a spreadsheet which outlines the required dates for these publishers/

You will notice that most of those with an empty first published date should indeed not have one, because they have only published private files.

For the 'Date is not valid' instances, I have provided that first published dates that we hold on our internal CRM system.

Let me know if you have any queries!

Thanks,
Josh
[First Published Updated.xlsx](https://github.com/IATI/ckanext-iati/files/3670589/First.Published.Updated.xlsx)
